### PR TITLE
Vite testing

### PR DESCRIPTION
**Pull request by simonachkar**

### CODE REVIEW

**Comment by TPNAYAK** [comment]
**Test Instructions:**

To test this PR:

1.  Clone the repository and navigate to the project directory.
2.  Run `npm install` to install the dependencies.
3.  Run `npm run dev` to start the development server.
4.  Open the application in your browser at `http://localhost:5173`.
5.  Verify that the main heading and Canada's flag are displayed.
6.  Check that the list of provinces is displayed correctly.
7.  Click on the 'Territories' link and ensure the list updates to show the territories.
8.  Click on the 'Provinces' link and ensure it returns to the province list.
9.  Interact with the counter component. Verify that the counter increments when the '+' button is clicked.
10. Verify that the counter resets to zero when the 'Reset' button is clicked.
11. Verify that the counter value persists after refreshing the page by leveraging local storage.
12. Test if the Show/Hide button works properly.

**Comment by TPNAYAK** [comment]
**Review Summary:**

This PR migrates the project from Create React App to Vite and adds a simple counter component with local storage. It also includes some updates to the README with new installation and usage instructions.

File: README.md [hunk 1]
```diff
@@ -1,40 +1,35 @@
-# Hello Canada....
+# Hello Canada
 
-Hello Canada, is a simple demo app that showcases the provinces and territories of Canada. It was built using Create React App, a popular tool for bootstrapping React applications.
+"Hello Canada" is a simple demo app showcasing the provinces and territories of Canada. It's built with **React** and powered by **Vite** for a fast development experience.
 
 ## Getting Started
 
-To get started with Hello Canada, follow these steps:
```
Comment (TPNAYAK) [inline README.md:7]: Consider adding the repository URL after the clone command for discoverability of the project on github.

File: src/components/counter.jsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+const LOCAL_STORAGE_KEY = 'counterValue';
+
+function Counter() {
+  const [count, setCount] = useState(() => {
+    const storedCount = localStorage.getItem(LOCAL_STORAGE_KEY);
+    return storedCount ? parseInt(storedCount, 10) : 0;
+  });
+
+  // New state to manage visibility
+  const [isVisible, setIsVisible] = useState(false);
```
Comment (TPNAYAK) [inline src/components/counter.jsx:13]: Consider adding a try/catch block to handle localStorage exceptions or use a library. Also you might want to use JSON parse and stringify when using `localStorage`.

File: README.md [hunk 3]
```diff
@@ -1,40 +1,35 @@
...
+    cd hello-canada
+    ```
+2.  **Install dependencies:**
+    ```bash
+    npm install
+    # or yarn install
+    ```
+3.  **Start the development server:**
+    ```bash
+    npm run dev
```
Comment (TPNAYAK) [inline README.md:19]: Consider adding info about yarn users as well.

File: src/App.jsx [hunk 4]
```diff
@@ -20,13 +22,14 @@ function App() {
 
   return (
     <div className="App">
-      <h1>Hello Canada</h1>
+      <h1>Hello Canada (A demo React App)</h1>
       <img
         alt="Canada's Flag"
         width={125}
         src="https://upload.wikimedia.org/wikipedia/commons/d/d9/Flag_of_Canada_%28Pantone%29.svg"
```
Comment (TPNAYAK) [inline src/App.jsx:29]: Recommend adding a fallback in case of API failure, which would handle cases where data is not loaded correctly, such as 'no provinces and territories'. Additionally, use `try/catch` block or `Promise.catch` to handle any errors.

File: README.md [hunk 5]
```diff
@@ -1,40 +1,35 @@
...
+    # or yarn dev
+    ```
+    Your app will be ready at `http://localhost:5173/`.
 
-1. Clone the repository to your local machine.
```
Comment (TPNAYAK) [inline README.md:9]: Consider creating a separate section for API documentation and usage.

File: src/components/counter.jsx [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  };
+
+  return (
+    <div className="counter-wrapper">
+      <button onClick={toggleVisibility} className="show-counter-btn">
+        {isVisible ? 'Hide Counter' : 'Show Counter'}
+      </button>
+
+      {/* Conditionally render the counter content based on isVisible state */}
+      {isVisible && (
```
Comment (TPNAYAK) [inline src/components/counter.jsx:38]: Consider externalizing the styles rather than adding them inside the component logic.

File: src/components/counter.jsx [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    const storedCount = localStorage.getItem(LOCAL_STORAGE_KEY);
+    return storedCount ? parseInt(storedCount, 10) : 0;
+  });
+
+  // New state to manage visibility
+  const [isVisible, setIsVisible] = useState(false);
+
+  useEffect(() => {
+    localStorage.setItem(LOCAL_STORAGE_KEY, count.toString());
+  }, [count]);
```
Comment (TPNAYAK) [inline src/components/counter.jsx:17]: Recommend using descriptive names for function variables for clarity.

File: README.md [hunk 8]
```diff
@@ -1,40 +1,35 @@
...
-2. Navigate to the project directory.
-3. Run `npm install` to install the project dependencies.
-4. Run `npm start` to start the development server.
-5. Open [http://localhost:3000](http://localhost:3000) in your browser to view the app.
```
Comment (TPNAYAK) [inline README.md:13]: If the application is not deployed anywhere, consider removing this section or provide instructions for local usage of json-server.

File: src/App.jsx [hunk 9]
```diff
@@ -1,6 +1,8 @@
 import React, { useEffect, useState } from "react";
 import List from "./components/list";
+import Counter from "./components/counter"; // Import the new Counter component
```
Comment (TPNAYAK) [inline src/App.jsx:3]: It is recommended to create separate folders for the components and css files rather than placing all of them in the `src` directory.

File: src/components/counter.jsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
+import { useState, useEffect } from 'react';
+import './counter.css'; 
+
```
Comment (TPNAYAK) [inline src/components/counter.jsx:3]: Is it a requirement to read data from localStorage inside the component and initialize the component count value with it? If it is only a demonstration, then please ignore this comment. In general, try not to use browser apis on server rendered components or outside event handlers. You may also want to use a library such as `usehooks-ts` instead.

File: src/App.jsx [hunk 11]
```diff
@@ -20,13 +22,14 @@ function App() {
 
   return (
     <div className="App">
-      <h1>Hello Canada</h1>
+      <h1>Hello Canada (A demo React App)</h1>
       <img
```
Comment (TPNAYAK) [inline src/App.jsx:26]: Consider storing the api url in .env files or a config file and accessing it via `process.env`

File: vite.config.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import { defineConfig } from "vite";
+import react from "@vitejs/plugin-react";
+
+export default defineConfig({
+  plugins: [react()],
+  build: {
+    outDir: "build",
```
Comment (TPNAYAK) [inline vite.config.js:7]: Consider adding code coverage support to your test config.

File: src/App.test.jsx [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    // 1. Load provinces initially
+    await screen.findByText('Ontario - Toronto');
+
+    // 2. Switch to territories
+    mockFetch.mockResolvedValueOnce({
+      ok: true,
+      json: () => Promise.resolve([{ name: 'Nunavut', capital: 'Iqaluit' }]),
+    });
+    fireEvent.click(screen.getByText('Territories'));
+    await screen.findByText('Nunavut - Iqaluit');
```
Comment (TPNAYAK) [inline src/App.test.jsx:112]: Consider adding more assertions and tests to the components. Use edge cases where possible. If there is logic for a specific use case in the component, make sure to add a test case for that use case.

File: package.json [hunk 14]
```diff
@@ -8,14 +8,13 @@
...
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
-    "react-scripts": "5.0.1",
     "web-vitals": "^2.1.4"
   },
   "scripts": {
-    "start": "react-scripts start",
-    "build": "react-scripts build",
-    "test": "react-scripts test",
-    "eject": "react-scripts eject"
```
Comment (TPNAYAK) [inline package.json:18]: Recommend adding a description, repository, and author information to the package.json file for better project metadata.
